okhttp/src/main/java/okhttp3/OkHttpClient.java

```diff
@@ -844,24 +846,30 @@ public Builder dispatcher(Dispatcher dispatcher) {
      */
     public Builder protocols(List<Protocol> protocols) {
       // Create a private copy of the list.
-      protocols = new ArrayList<>(protocols);
-
-      // Validate that the list has everything we require and nothing we forbid.
-      if (!protocols.contains(Protocol.HTTP_1_1)) {
-        throw new IllegalArgumentException("protocols doesn't contain http/1.1: " + protocols);
+      final Set<Protocol> uniqueProtocols = new HashSet<>(protocols);
+
+      if (uniqueProtocols.contains(Protocol.H2C) && uniqueProtocols.size() > 1) {
+        // when using h2c prior knowledge, no other protocol should be supported.
+        throw new IllegalArgumentException("protocols containing h2c cannot use other protocols:"
+                + uniqueProtocols);
+      } else if (!uniqueProtocols.contains(Protocol.HTTP_1_1)) {
+        // Validate that the list has everything we require and nothing we forbid.
+        throw new IllegalArgumentException("protocols doesn't contain http/1.1: "
+                + uniqueProtocols);
       }
-      if (protocols.contains(Protocol.HTTP_1_0)) {
-        throw new IllegalArgumentException("protocols must not contain http/1.0: " + protocols);
+      if (uniqueProtocols.contains(Protocol.HTTP_1_0)) {
+        throw new IllegalArgumentException("protocols must not contain http/1.0: "
+                + uniqueProtocols);
       }
-      if (protocols.contains(null)) {
+      if (uniqueProtocols.contains(null)) {
         throw new IllegalArgumentException("protocols must not contain null");
       }
 
       // Remove protocols that we no longer support.
-      protocols.remove(Protocol.SPDY_3);
+      uniqueProtocols.remove(Protocol.SPDY_3);
 
       // Assign as an unmodifiable list. This is effectively immutable.
-      this.protocols = Collections.unmodifiableList(protocols);
+      this.protocols = Collections.unmodifiableList(new ArrayList<>(uniqueProtocols));
```

Comment:
I think changing the selected protocol because of hashing could affect the ALPN selection IIRC.  Not so much an issue without SPDY, but undesirable.

Reply:
Yeah. I think that change went a little too far. Going to revert back to using just the list and suffer the minimal performance cost.